### DISCUSSION TITLE
Can SPARQL Graph generate aggregate queries?

### DISCUSSION THREAD

**Discussion by Robert-Adelard** [ge-high-assurance/RACK#543] [Q&A] [answered]
Can SPARQL Graph be used to construct aggregate queries, or do you have to write the necessary SPARQL queries by hand?

https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#aggregates

For example, could I construct a query that counted how many test results / test executions I had for each test case?

**Comment by cuddihyge** [comment, marked as answer]
GROUP BY, HAVING, and functions like SUM() are not currently supported in nodegroups.
These are on the backlog.
I think these are excellent candidates for improved functionality in the near future.  @kityansiu @AbhaMoitra and can discuss.

**Reply by Robert-Adelard** [reply]
Thanks - I agree that these would be useful features to have. I'd also like to see a MINUS operator that could be used to subtract one nodegroup from another - see #515.

This isn't just theoretical - we want to construct queries of this form and are currently unable to do so.

**Reply by davearcher** [reply]
Is there a broader expected set of SQL-similar operators that you'd like to see, @Robert-Adelard ? And to what degree are recursive queries or other sub-queries concretely known to be of interest?

**Reply by Robert-Adelard** [reply]
I'd certainly like to see support for the full range of queries that can be expressed in SPARQL - the issue is how to express them naturally with a graphical notation. And if nodegroups are to be a useful abstraction, I think there needs to be a way of composing nodegroups and treating them as composite entities. So it would be good to have the ability to construct queries from sub-queries. I'm not sure about recursion, but transitive closure over a relation can be useful.
